File: eummd/src/eummd.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <math.h>
#include <random>

#include "medianHeuristic.h"

/**
 * Merge two already sorted vectors into a third vector. 
 *
 * Note that it is a template function and the values are of type
 * Comparable (although this is hidden in vectype).
 * Not exposed.
 *
 * @param A First vector, sorted.
 * 
 * @param B Second vector, sorted.
 * 
 * @return C The merged vector, the sorted vector which is the merge of two other sorted vectors.
 */
std::vector<double> mergeTwoAlreadySorted(std::vector<double>& A, 
                                          std::vector<double>& B){

    // allocate space for C
    std::vector<double>::size_type n = A.size() + B.size();
    std::vector<double> C(n);

    //iterators
    std::vector<double>::const_iterator a_iter = A.begin();
    std::vector<double>::const_iterator b_iter = B.begin();
    //iterator for C is not constant, because assigning values form A and B
    std::vector<double>::iterator c_iter = C.begin();
    
    //Filling C while comparing A and B
    while (  (a_iter != A.end()) && (b_iter != B.end()) && (c_iter != C.end()) ){
        if ( (*a_iter) < (*b_iter) ){
            *c_iter = *a_iter;
            ++a_iter;
        } else {
            *c_iter = *b_iter;
            ++b_iter;
        }
        ++c_iter;
    }

    //now fill in the rest
    //only a *maximum* of one of these two while loops will run
    //at least one of the iterators reached the end in the previous while
    while ( (a_iter != A.end()) && (c_iter != C.end()) ){
        *c_iter = *a_iter;
        ++a_iter;
        ++c_iter;
    }
    while ( (b_iter != B.end()) && (c_iter != C.end()) ){
        *c_iter = *b_iter;
        ++b_iter;
        ++c_iter;
    }

    //return the merged vector
    return C;
}




/* Sorted sum of distances with Laplacian Kernel
 *
 * Compute the sorted sum of distances for a vector, 
 * while using the Laplacian kernel.
 *
 * @param Z Some kind of vector. 
 *
 * @param beta The parameter for the Laplacian kernel.
 *
 * @see sortedSumofDist
 *
 * @return The sorted sum of distances using the Laplacian 
 *         kernel with parameter beta.
 */
    double lapKernSSD(const std::vector<double>& Z, double beta){
        // row sum for first row
        double R = 0.0;
        // row sum for second row
        double S = 0.0;

        // not starting for loop from 0, but from 1
        // note that not i-1 but is i in R += i * D
        // because of indexing from 0, and running to n-1
        // so create iterator and increase by 1
        std::vector<double>::const_iterator it = Z.begin();
        double Z_i = *it;
        double Z_i_1 = *it;
        ++it;
        for (; it != Z.end(); ++it){
            Z_i = *it;
            R = (R+1) * std::exp(  -beta * (Z_i - Z_i_1)  );
            S += R;
            Z_i_1 = Z_i;
        }
        return S;
    }


/* FastMMD calculation
 * 
 * Computes MMD statistic from 
 *
 * @param Zstart Constant iterator at start of Z vector.
 *
 * @param n1 Size of first sample.
 *
 * @param n2 Size of second sample.
 *
 * @param permstart Constant iterator at start of perm bool vector.
 *
 * @param beta The parameter/bandwidth for the kernel.
 *
 * @details Assumes Z is sorted.
 *
 * @see lapKernSSD
 *
 * @return MMD statistic.
 */
double compute_eummd_faster(std::vector<double>::const_iterator Zstart,
                           const std::vector<double>::size_type n1, 
                           const std::vector<double>::size_type n2, 
                           std::vector<bool>::const_iterator permstart, 
                           std::vector<bool>::const_iterator permend, 
                           const double T4, 
                           const double beta){

    // MMD_u statistic
    double MMD_u = 0.0;

    // initialise X and Y vectors from Z and reserve space for push_back
    std::vector<double> X; 
    X.reserve(n1);
    std::vector<double> Y; 
    Y.reserve(n2);

    //define iterators
    std::vector<double>::const_iterator itZ = Zstart;
    std::vector<double>::const_iterator Zend = Zstart + n1+n2 ;

    std::vector<bool>::const_iterator itperm = permstart;
    // permend already defined

    // run through permutation (filter); if true, goes to X, else goes to Y
    // code is a bit longer, but verctor only run through once now
    // Could also filter Z into X and Y using std::copy_if
    // https://stackoverflow.com/questions/21204676/modern-way-to-filter-stl-container
    for ( ; (itperm != permend) && (itZ != Zend); ++itperm, ++itZ){
        if (*itperm==true){
           X.push_back(*itZ);
        } else {
           Y.push_back(*itZ);
        }
    }


    // compute other terms; two O(n) steps
    double T1 = lapKernSSD(X, beta);
    double T2 = lapKernSSD(Y, beta);
    double T3 = T4 - T1 - T2;

    //now compute MMD - need to multiply T1 and T2 by 2
    double c1 = 0;
    if (n1 > 1){
        c1 =  2.0 / ( n1 * (n1 - 1) );
    }
    double c2 = 0;
    if (n2 > 1){
        c2 =  2.0 / ( n2 * (n2 - 1) );
    }
    double c3 = -2.0 / (n1 * n2);

    MMD_u = c1*T1 + c2*T2 + c3*T3;
    return(MMD_u);
}


std::vector<double> cpp_eummd_pval_faster(std::vector<double> X, 
                                          std::vector<double> Y, 
                                         double beta, int numperm, int seednum, 
                                         int twosided, int boundedminpval){


    // return vector is first pval, then statistic
    std::vector<double> returnvec;

    // sort each set
    // O(n1 log n1 + n2 log n2) = O(n log n)
    std::sort(X.begin(), X.end());
    std::sort(Y.begin(), Y.end());

    // merge two sorted vectors into Z
    // O(n)
    std::vector<double> Z = mergeTwoAlreadySorted(X, Y);

    // if beta is NOT a positive number, then compute using median heuristic
    // O(n log n)
    if ( !(beta > 0) ){
        // // making a copy, just in case
        // std::vector<double> Zcopy;
        // Zcopy.insert(Zcopy.end(), Zstart, Zend);
        double medianDiff = medianHeuristicAlreadySorted(Z);
        // beta is set to 1/ median difference
        beta = 1.0 / medianDiff;
    }

    // the triangular sum of the whole set; used in all calculations 
    // O(n)
    double T4 = lapKernSSD(Z, beta);

    const std::vector<double>::size_type n1 = X.size();
    const std::vector<double>::size_type n2 = Y.size();

    // create permutation
    std::vector<bool> Xperm(n1, true);
    std::vector<bool> Yperm(n2, false);
    // concatenating permutations into perm
    std::vector<bool> perm(Xperm);
    perm.insert(perm.end(), Yperm.begin(), Yperm.end());

    //make new vector, where X at front, Y at end
    //this is different from Z, which is completely sorted
    //This is only for MMDstar; just to extract X and Y properly
    //otherwise need to find permutation of 0s and 1s that 
    //will extract X and Y from Z correctly in compute_eummd_faster
    //other option is to extract X and Y in this method
    std::vector<double> XY(X);
    XY.insert( XY.end(), Y.begin(), Y.end() );
    std::vector<double>::const_iterator XYstart = XY.begin();
    double MMDstar = compute_eummd_faster(XYstart, n1, n2, 
                                         perm.begin(), perm.end(), 
                                         T4, beta);

    // initiate random generator
    std::random_device rd;
    std::mt19937 g;
    // set the seed, if greater than 0
    if (seednum > 0){
        g.seed(seednum);
    } else {
        g.seed(rd());
    }

    double MMDperm = 0.0;
    //counting the value already computed
    int MMD_count_below = 1;

    //run the permutations
    for (int i=0; i < numperm; ++i){
        //shuffle permutation and compute MMD
        std::shuffle(perm.begin(), perm.end(), g);
        // pass Z, which is sorted
        MMDperm = compute_eummd_faster(Z.begin(), n1, n2, 
                                      perm.begin(), perm.end(), 
                                      T4, beta);

        if (MMDperm < MMDstar){
            ++MMD_count_below;
        }
    }

    //now return the threshold; this will be the pvalue
    double pval = MMD_count_below / (numperm + 1.0);

    //make one-sided
    //twosided = 0 means false
    //twosided = 1 means true
    if (twosided==1){
        //make two sided
        pval =  1 - std::abs(1 - 2*pval) ;
    } else {
        //one sided; counting below, so must make 
        // large values into small
        pval = 1 - pval;
    }

    // quick check for minimum possible p-value; avoids pval=0
    // 1 / 2(numperm+1)
    if (boundedminpval==1){
        double pmin = 0.5 / (numperm+1.0);
        if (pval < pmin)
            pval = pmin;
    }

    // return vector is first pval, then statistic
    returnvec.push_back(pval);
    returnvec.push_back(MMDstar);
    returnvec.push_back(beta);
    return returnvec;
} // end of cpp_eummd_pval_faster



// only called once by cpp_eummd
// may compute beta if < 0
std::vector<double> eummd_ptr_alt(std::vector<double>::const_iterator Zstart, 
                const std::vector<double>::size_type nX, 
                const std::vector<double>::size_type nY, 
                double beta){

    // return vector is first statistic, then beta 
    std::vector<double> returnvec;

    std::vector<double>::const_iterator Xstart = Zstart;
    std::vector<double>::const_iterator Xend = Xstart + nX;
    std::vector<double>::const_iterator Ystart = Xend;
    std::vector<double>::const_iterator Yend = Ystart + nY;

    // create **deep** copies of vectors; we will work on these.
    std::vector<double> X(Xstart, Xend);
    std::vector<double> Y(Ystart, Yend);

    //first, get the lengths of the vectors
    std::vector<double>::size_type n1 = X.size();
    std::vector<double>::size_type n2 = Y.size();

    // second, sort each set
    std::sort(X.begin(), X.end());
    std::sort(Y.begin(), Y.end());

    //third, merge the two sorted vectors into a third
    std::vector<double> Z = mergeTwoAlreadySorted(X, Y);

    //(optional) compute median heuristic
    // if beta is NOT a positive number, then compute using median heuristic
    // O(n log n)
    if ( !(beta > 0) ){
        // // making a copy, just in case
        // std::vector<double> Zcopy;
        // Zcopy.insert(Zcopy.end(), Zstart, Zend);
        double medianDiff = medianHeuristicAlreadySorted(Z);
        // beta is set to 1/ median difference
        beta = 1.0 / medianDiff;
    }

    //fourth, get the first two sorted sums
    double T1 = lapKernSSD(X, beta);
    double T2 = lapKernSSD(Y, beta);

    //fifth, compute the triangular sum for all the values
    double T4 = lapKernSSD(Z, beta);

    //sixth, now do scalar computations to get the solution
    double T3 = T4 - T1 - T2;

    //now compute MMD - need to multiple T1 and T2 by 2
    double c1 = 0;
    if (n1 > 1){
        c1 =  2.0 / ( n1 * (n1 - 1) );
    }
    double c2 = 0;
    if (n2 > 1){
        c2 =  2.0 / ( n2 * (n2 - 1) );
    }
    double c3 = -2.0 / (n1 * n2);

    double MMD_u = c1*T1 + c2*T2 + c3*T3;

    // push statistic and beta to returnvec
    returnvec.push_back(MMD_u);
    returnvec.push_back(beta);
    return returnvec;
} //end eummd_ptr_alt



std::vector<double> cpp_eummd(std::vector<double> X, 
                              std::vector<double> Y, double beta){
     
    std::vector<double> Z(X); 
    // then insert Y into it
    Z.insert(Z.end(), Y.begin(), Y.end());

    return eummd_ptr_alt(Z.begin(), X.size(), Y.size(), beta);
}

File: eummd/R/eummd.R
#' euMMD: Efficient Univariate Maximum Mean Discrepancy
#'
#' Computes the maximum mean discrepancy statistic with the Laplacian kernel. 
#' Suitable only for univariate data. Computing the statistic alone
#' for \eqn{n} observations is \eqn{O(n \log n)}, and computing the 
#' p-value for \eqn{L} permutations is \eqn{O(n \log n + Ln)}.
#' 
#' @param x Univariate vector of observations in first sample.
#' 
#' @param y Univariate vector of observations in second sample.
#' 
#' @param beta kernel parameter. Must be positive; if not, computes
#'             median heuristic in quadratic time. Default value
#'             is \code{-0.1}, which will force median heuristic to be used.
#' 
#' @param pval Boolean for whether to compute p-value or not. 
#' 
#' @param numperm Number of permutations. Default is \code{200}.
#'
#' @param seednum Seed number for generating permutations. Default is \code{0}, 
#'                which means seed is set randomly. For values larger than 
#'                \code{0}, results will be reproducible.
#' 
#' @param alternative A character string specifying the alternative hypothesis,
#'                    which must be either \code{"greater"} (default) or
#'                    \code{"two.sided"}. In Gretton et al., the 
#'                    MMD test statistic is specified so that if it is 
#'                    significantly larger than zero, then the null hypothesis
#'                    that the two samples come from the same distribution 
#'                    should be rejected. For this reason, \code{"greater"}
#'                    is recommended. The test will still work 
#'                    in many cases with \code{"two.sided"} specified, but this
#'                    could lead to problems in certain cases.
#'
#' @param allowzeropval A boolean, specifying whether we will allow zero 
#'                      p-values or not. Default is \code{FALSE}; then 
#'                      a threshold of \code{0.5 / (numperm+1)} is used, 
#'                      and if the computed p-value is less than this
#'                      threshold, it is then set to be this value.
#'                      this avoids the possibility of zero p-values.
#'
#' @details If the total number of observations in both samples is \code{n}, 
#'          first sort combined sample in \eqn{O(n \log n)} before remaining
#'          steps are linear in \code{n}.
#' 
#'    If \code{beta} is not a positive value, 
#'    median difference is computed as follows:
#'          
#'    \deqn{ m = \textnormal{median} \{ || x_i - x_j ||_1; \,\, i>j, \,\, 
#'           i=1, 2,\dots, n+m,\,\,\textnormal{ and } j=1, 2,\dots, i-1 \}, }
#'    
#'    where \eqn{ || x_i - x_j ||_1} is the 1-norm, and 
#'    so if the data are univariate then
#'    
#'    \deqn{ || x_i - x_j ||_1 = |x_{i} - x_{j}|. }
#'    
#'    and finally median heuristic is \code{beta = 1/m}.
#'    This can be computed in \eqn{O(n \log n )} time
#'    using the algorithms of Johnson and Mizoguchi (1978) 
#'    and Croux and Rousseuw (1992); see \code{mediandiff}
#'    for references.
#'    
#'    The Laplacian kernel \eqn{k} is defined as 
#'    
#'    \deqn{ k(x,y) = \exp [ -\beta ||x - y||_1  ]. }
#'
#'    The random seed is set for \code{std::mt19937} and \code{std::shuffle} 
#'    in C++.
#'
#'
#'
#' @return A list with the following elements:
#'         \describe{
#'             \item{\code{pval}}{The p-value of the test, if it is  
#'                                computed (\code{pval=TRUE}). Otherwise, 
#'                                it is set to \code{NA}.}
#'             \item{\code{stat}}{The statistic of the test, which
#'                                is always computed. }
#'             \item{\code{beta}}{The kernel parameter used in the test.
#'                                If \code{beta} was not initialised or
#'                                negative, this will be the median heuristic
#'                                value.}
#'          }
#' 
#' @references
#'    Bodenham, D. A., and Kawahara, Y. (2023)
#'    "euMMD: efficiently computing the MMD two-sample test statistic for 
#'    univariate data." Statistics and Computing 33.5 (2023): 110.
#' 
#'    Croux, C. and Rousseeuw, P. J. (1992), 
#'    "Time-Efficient Algorithms for Two Highly Robust Estimators of Scale"
#'    In Computational Statistics: Volume 1: Proceedings of the 10th 
#'    Symposium on Computational Statistics (pp. 411-428).
#'
#'    Johnson, D.B., and Mizoguchi, T. (1978), 
#'    "Selecting the Kth Element in X + Y and X_1 + X_2 + ... + X_m", 
#'    SIAM Journal of Computing, 7, 147-153.
#' 
#' @seealso \code{mediandiff}
#'
#' @examples 
#' x <- c(7.1, 1.2, 4.3, 0.4)
#' y <- c(5.5, 2.6, 8.7)
#' #setting the kernel parameter to be 0.1; setting seed=1 for reproducibility
#' mmd_list <- eummd(x, y, beta=0.1, seednum=1)
#'
#' #now using median heuristic (default)
#' mmd_list <- eummd(x, y, seednum=1)
#'
#' #now not computing the p-value, only the statistic
#' mmd_list <- eummd(x, y, pval=FALSE, seednum=1)
#' \donttest{
#' #now using a larger number of permutations 
#' mmd_list <- eummd(x, y, numperm=1000, seednum=1)
#' }
#'
#'
#' @export
eummd <- function(x, y, beta=-0.1, pval=TRUE, numperm=200, seednum=0, 
                  alternative=c("greater", "two.sided"), allowzeropval=FALSE){

    # check vectors are numeric
    if ( !(is.numeric(x)) || !(is.vector(x)) ){
        stop("x needs to be a numeric vector.")
    }
    if ( !(is.numeric(y)) || !(is.vector(y))){
        stop("y needs to be a numeric vector.")
    }

    # if alternative is 'greater', default
    alternative <- alternative[1]
    if ( (alternative != "greater") && (alternative != "two.sided") ){
        stop("alternative needs to be either 'greater' or 'two.sided'.")
    }

    # two sided is false by default; false is 0
    twosided <- 0
    if (alternative == "two.sided")
        twosided <- 1

    # boundminpval is true by default; true is 1, false is 0
    boundminpval <- 1
    if (allowzeropval==TRUE)
        boundminpval <- 0


    mmdList <- list()
    if (pval){
        mmdList <- eummd_pval_Rcpp(x, y, beta, numperm, seednum, 
                                   twosided, boundminpval)
    } else {
        mmdList <- eummd_Rcpp(x, y, beta)
    }

    # check pval; if no pval, functions return -1, then changes to NA
    if (mmdList$pval < 0){
        mmdList$pval <- NA
    }
    return(mmdList)
}

